File: py_solutions_21-30/Euler_27.py
# Find the product of the coefficients, a and b, for the quadratic 
# expression that produces the maximum number of primes for consecutive values of n, starting with n = 0.

from __future__ import print_function

import math
import timeit
import itertools

try:
    range = xrange
except NameError:
    pass

start = timeit.default_timer()

def is_prime(x):
    if x == 2:
        return True
    if x % 2 == 0 or x == 1:
        return False
    for i in range(3, int(math.sqrt(x)) + 1, 2):
        if x % i == 0:
            return False
    return True


def euler_27():
    max_primes = 0
    quadratic = lambda x, y, z: x**2 + (x * y) + z
    for i, v in itertools.product(range(-999, 1000), range(-999, 1000)):
        primes = 0
        q = quadratic(primes, v, i) 
        while is_prime(abs(q)):
            primes += 1
            q = quadratic(primes, v, i)
        if primes > max_primes:
            max_primes = primes
            co_prod =  i * v
    return co_prod


print("Answer: {}".format(euler_27()))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format(stop - start))

File: Perl/perl_solutions_21-30/Euler_27.pl
use strict;
use warnings;
use Math::Complex;
use Time::HiRes qw( clock );


sub is_prime 
{
    my ($n) = @_;
    if ($n == 2) {
        return 1;
    }
    if ($n == 1 || $n % 2 == 0) {
        return 0;
    }
    for my $i (3..int(sqrt($n)) + 1) {
        if ($n % $i == 0) {
            return 0;
        }
    }
    return 1;
}

sub quadratic_primes
{
    my $max = 0;
    my $co_prod = 0;
    for my $i (-999..999) {
        for my $j (-999..999) {
            my $primes = 0;
            my $q = $primes**2 + ($primes * $j) + $i;
            while (is_prime(abs($q))) {
                $primes++;
                $q = $primes**2 + ($primes * $j) + $i;
            }
            if ($primes > $max) {
                $max = $primes;
                $co_prod = $i * $j;
            }
        }
    }
    return $co_prod;
}
    

my $start = clock();
my $answer = quadratic_primes();
my $stop = clock();
print "Answer: $answer\n";
printf "Time: %.5f\n", $stop - $start;
